src/libtiled/tilelayer.cpp

```diff
@@ -199,57 +199,132 @@ void TileLayer::flip(FlipDirection direction)
 
     Q_ASSERT(direction == FlipHorizontally || direction == FlipVertically);
 
-    for (int y = 0; y < mHeight; ++y) {
-        for (int x = 0; x < mWidth; ++x) {
-            Cell &dest = newGrid[x + y * mWidth];
-            if (direction == FlipHorizontally) {
-                const Cell &source = cellAt(mWidth - x - 1, y);
-                dest = source;
-                dest.setFlippedHorizontally(!source.flippedHorizontally());
-            } else if (direction == FlipVertically) {
-                const Cell &source = cellAt(x, mHeight - y - 1);
-                dest = source;
-                dest.setFlippedVertically(!source.flippedVertically());
-            }
-        }
-    }
+    const Map* const globalMap = map();
+    const Map::Orientation currentOrientation = (globalMap ? globalMap->orientation() : Map::Unknown);
```

Comment:
Since we can't rely on the tile layer being part of a map in general, and because of the duplicated implementation, I think it would make sense to split this up into `flip` and `flipHexagonal`, `rotate` and `rotateHexagonal`. We can decide which version to call in `tilestamp.cpp`, where a null check on the map is not needed.